src/shared/sol-iio.c

```diff
@@ -0,0 +1,990 @@
+
+    channel->shift = 0;
+    len = sscanf(type, "%d/%d>>%d", &channel->bits, &channel->storagebits, &channel->shift);
+
+    free(original_type);
+
+    if (len < 0)
+        return false;
+
+    return true;
+}
+
+/* Some channels are named on the form  <type>[_x|_y|_z]
+ * This function shall return the name without <x|y|z> components.
+ * The form <type>[Y][_modifier] is also common (Y is a number), and this function
+ * remove numbers too, in an attempt to get the 'pure' name.
+ * TODO there are other esoteric combinations - for them, if we care about,
+ * we will probably need API calls to get the 'scale' and 'offset' file names */
+static char *
+channel_get_pure_name(struct sol_iio_channel *channel)
+{
+    size_t channel_name_len, i;
+    char *channel_pure_name, *channel_name_suffix, *original_channel_pure_name;
+    bool modified = false;
+
+    channel_name_len = strlen(channel->name);
+    if (channel_name_len > 2) {
+        channel_name_suffix = channel->name + channel_name_len - 2;
+        if (streq(channel_name_suffix,  "_x") ||
+            streq(channel_name_suffix,  "_y") || streq(channel_name_suffix,  "_z")) {
+
+            channel_pure_name = malloc(channel_name_len - 1);
+            strncpy(channel_pure_name, channel->name, channel_name_len - 2);
+            channel_pure_name[channel_name_len - 2] = '\0';
+            return channel_pure_name;
+        } else {
+            /* Recreate channel name without Y_ components (Y is a number).
+             * Idea is 's/[0-9]+//' */
+            channel_pure_name = calloc(1, channel_name_len + 1);
+            original_channel_pure_name = channel_pure_name;
+            for (i = 0; i < channel_name_len; i++) {
+                if (isalpha(channel->name[i]) || channel->name[i] == '-' || channel->name[i] == '_')
+                    *channel_pure_name++ = channel->name[i];
+                else if (isdigit(channel->name[i])) {
+                    modified = true;
+                    continue;
+                }
+            }
+            if (modified)
+                return original_channel_pure_name;
+            else
+                free(original_channel_pure_name);
+        }
+    }
+
+    return NULL;
+}
+
+static void
+channel_get_scale(struct sol_iio_channel *channel)
+{
+    char path[PATH_MAX + 1];
+    int len;
+    char *channel_pure_name = NULL;
+    struct sol_iio_device *device = channel->device;
+
+#define GET_SCALE(_name) \
+    do { \
+        len = snprintf(path, sizeof(path), CHANNEL_SCALE_PATH, device->device_id, \
+            _name); \
+        if (len > 0 && sol_util_read_file(path, "%lf", &channel->scale) > 0) { \
+            goto end; \
+        } \
+    } while (0)
+
+    GET_SCALE(channel->name);
+    /* No scale. If channel has x, y, z or Y component, look for scale file without it */
+    channel_pure_name = channel_get_pure_name(channel);
+    GET_SCALE(channel_pure_name);
```

Comment:
But macro `GET_SCALE` already frees it (actually `goto end`, where there's a free); maybe I could rewrite this code to be more clear.